Adding support for Sphere payment processing

bradymiller opened the pull request:
This supports (Sphere tool that was integrated is in parenthesis):
1. Payment capture from front office (Trustee Premier for MOTO)
2. Payment capture from patient portal (Trustee Premier for Ecomm)
3. void/credit of payments (Trustee API)

Main advantage of this approach is that all the sensitive info (cc number) is captured on Sphere's web server and not done on OpenEMR's server. Note there are 3 different Sphere tools that are utilizing, which are listed above, and is why there are several global credentials required for this. Note the front office uses a card reader (or the machine can enter in the cc number); btw, the machines for front office basically leverage native keyboard drivers so super easy to work (basically just focus the field on web page and then read card or enter data into machine).

Also added a generic mechanism to support encrypted hash in globals since needed a pin mechanism to ensure void/credit is authorized for user within OpenEMR itself. Using this for the `Sphere Void/Credit Confirmation PIN` global.

I considered a module (which is why the code is operating like a module by inserting the code via one liners), but was really awkward because of the current payment processing options (am also hoping this becomes the main payment processing that is used since the OpenEMR foundation will actually get funding when this payment system is used).

Pertinent globals:
- Connectors->Accept Credit Card transactions from Front Payments
- (added Sphere option) Connectors->Select Credit Card Payment Gateway->Gateway for Sphere Payments
- Connectors->Set Gateway to Production Mode (don't turn this on until testing mode is working)
- (new) Connectors->Connectors->Sphere Clinicfront (MOTO) Transaction CustID
- (new) Connectors->Sphere Clinicfront (MOTO) Transaction CustID License Key
- (new) Connectors->Sphere MOTO TC Link Password
- (new) Connectors->Sphere Patientfront (Ecomm) Transaction CustID
- (new) Connectors->Sphere Patientfront (Ecomm) Transaction CustID License Key
- (new) Connectors->Sphere Ecomm TC Link Password
- (new) Connectors->Sphere Void/Credit Confirmation PIN
- Portal->Allow Online Payments

Some things that also need to ensure for this to work:
- Need to be using https
- The global Connectors->Site Address Override needs to be correct

Posted screenshots below for all 3 workflows.

---

# Screenshots

## Screenshots of the front office

![image](https://user-images.githubusercontent.com/278968/141053047-9867a870-567b-4f53-aa7a-057ea8e81e45.png)

![image](https://user-images.githubusercontent.com/278968/141053097-7d0f6b3e-4b7b-4052-aa55-f38343700fa9.png)

![image](https://user-images.githubusercontent.com/278968/141053160-fa85c130-2e4e-4a16-b16f-d868297fae47.png)

![image](https://user-images.githubusercontent.com/278968/143786358-1d72b27d-c463-42b0-9b5e-d382b5e2affc.png)

![image](https://user-images.githubusercontent.com/278968/141053241-c92190d1-6b71-4efd-97f7-235aeb2ca78f.png)

---

## Screenshots for patient portal

![image](https://user-images.githubusercontent.com/278968/141053328-74a4c790-3764-4861-89ad-5a2e883b7e3a.png)

![image](https://user-images.githubusercontent.com/278968/141053471-1265e61b-c519-4a3a-9fd5-7cba91ec5625.png)

![image](https://user-images.githubusercontent.com/278968/141053498-c4b13af9-8f8f-4540-a506-93e7a5d38711.png)

![image](https://user-images.githubusercontent.com/278968/141053535-e095120b-ac53-4fa1-819a-a668c6d89b0c.png)

---

## Screenshots for void/credit

![image](https://user-images.githubusercontent.com/278968/143786421-dc7d08d3-b436-499d-ab8f-9bbd6851fecf.png)

![image](https://user-images.githubusercontent.com/278968/143786478-a14021d0-8a46-445d-aec8-d07c2476a3a7.png)

![image](https://user-images.githubusercontent.com/278968/143786499-9044d02d-1608-425a-bbba-8f843191d616.png)

![image](https://user-images.githubusercontent.com/278968/143786509-e7fbdf22-88c0-4289-83b0-38cfaf52e659.png)

![image](https://user-images.githubusercontent.com/278968/143786540-aff8fd7a-a233-4930-bbfe-f03248adf27d.png)

---

bradymiller commented on sphere/process_response.php:
```diff
@@ -0,0 +1,83 @@
+
+    if ($_GET['front'] == 'patient') {
+        // in future, will always make this blank in patient front since should not be displayed in patient front
+        $description = (!empty($_POST['description'])) ? ' (' .  $_POST['description'] . ')' : '';
```
oops, forgot to do this

adunsulag commented on src/PaymentProcessing/SpherePayment.php:
```diff
@@ -0,0 +1,257 @@
+                let oForm = document.forms['payment-form'];
+                oForm.elements['mode'].value = 'Sphere';
+
+                let inv_values = JSON.stringify(getFormObj('invoiceForm'));
```
Do you care if someone inserts a javascript value that can't be stringified here?  I usually put these in try catch blocks.

adunsulag commented on src/PaymentProcessing/SpherePayment.php:
```diff
@@ -0,0 +1,257 @@
+                    return response.text();
+                }).then(function(data) {
+                    if(data !== 'ok') {
+                        alert(data);
```
I'd change this to be
```
throw data;
```
so it goes to the same catch block.

adunsulag commented on src/PaymentProcessing/SpherePayment.php:
```diff
@@ -0,0 +1,257 @@
+                    alert(chargeMsg);
+                    window.location.reload(false);
+                }).catch(function(error) {
+                    alert(error)
```
Do you not want to do a reload here? in the above then() clause if you hit an error you still reload the page.

adunsulag commented on src/PaymentProcessing/SpherePayment.php:
```diff
@@ -0,0 +1,257 @@
+    {
+        return "
+            function sphereSuccess(authCode, transId) {
+                $('#check_number').val(transId);
```
document.getElementById would be better here.

adunsulag commented on src/PaymentProcessing/SpherePayment.php:
```diff
@@ -0,0 +1,257 @@
+                $('#check_number').val(transId);
+                alert(" . xlj('Success') . " + '\\n' + 'Auth: ' + authCode + ' TransId: ' + transId);
+                // core payment process
+                $(\"[name='form_save']\").click();
```
You can use here:
```
document.querySelector("[name='form_save']").click();
```
or
```
document.querySelector("[name='form_save']").dispatchEvent(new Event('click'));
```

adunsulag commented on src/PaymentProcessing/SpherePayment.php:
```diff
@@ -0,0 +1,257 @@
+    {
+        return "
+            function sphereNotSuccess(message) {
+                alert(message);
```
Are we logging these messages anywhere?  Or does Sphere backend log these messages for the provider/IT go and look them up?  

I've always found that good payment processing contains a log of the errors so you can quickly fix and resolve issues.

adunsulag commented on src/PaymentProcessing/SpherePayment.php:
```diff
@@ -0,0 +1,257 @@
+                return result;
+            }
+            function paySphere() {
+                let total = $(\"[name='form_paytotal']\").val();
```
more jQuery that can be changed to use vanilla js.

adunsulag commented on src/PaymentProcessing/SpherePayment.php:
```diff
@@ -0,0 +1,257 @@
+                }
+
+                let responseUrl = " . js_escape($this->serverSite) . " + '/sphere/initial_response.php';
+                let cancelUrl = " . js_escape($this->serverSite) . " + '/sphere/initial_response.php?cancel=cancel&front=' + " . js_escape($this->front) . " + '&patient_id_cc=' + " . js_escape($this->patientIdCc) . " + '&csrf_token=' + " . js_escape(CsrfUtils::collectCsrfToken('sphere')) .  ";
```
Do we need to add any HMAC signature to make sure the variables were not modified / tampered with here? If another javascript was on this page someone could leave the CSRF alone but modify the other variables.  Maybe I'm being too paranoid here?

adunsulag commented on src/PaymentProcessing/SpherePayment.php:
```diff
@@ -0,0 +1,257 @@
+     */
+    public static function sphereAuditLog(int $pid, int $success, array $auditData, ?string $ticket = null, ?string $transactionId = null, ?string $actionName = null): void
+    {
+        $cryptoGen = new CryptoGen();
```
So this would be a good place if the transaction fails to store the results in the audit log (see javascript errors above).

adunsulag commented on sphere/initial_response.php:
```diff
@@ -0,0 +1,53 @@
+ * @license   https://github.com/openemr/openemr/blob/master/LICENSE GNU General Public License 3
+ */
+
+if (!empty($_GET['cancel']) && ($_GET['cancel'] == 'cancel')) {
```
Does Sphere have any kind of signature validation for you to validate that the request originated from them?  We probably only want to process requests if they originated from sphere.

adunsulag reviewed (commented):
Hey @bradymiller Did an initial pass through here.  I haven't pulled it to play around with it but just some thoughts here.

adunsulag commented:
@bradymiller BTW this might make a great live demonstration of turning code into a module for our Nov 20th training.  You wrote this pretty well and it will go over fairly quickly into a module.  I may go ahead and do that next week.

bradymiller commented:
Just add a payment processing "report" where can audit things and where plan to support void/credit feature. As everything else, is still a work in progress. Next goal is to get void/credit working.

bradymiller commented:
@adunsulag , Thanks for reviewing the code! All very good points and will begin to address them after get void/credit working.

bradymiller commented:
btw @adunsulag , definitely feel free to modularize this if will be a good example for the Nov 20th training. Even though it's not complete, whatever you end up doing to make it a module, I can integrate it in future.

bradymiller commented on sphere/process_response.php:
```diff
@@ -0,0 +1,82 @@
+use OpenEMR\Common\Csrf\CsrfUtils;
+use OpenEMR\Core\Header;
+use OpenEMR\PaymentProcessing\PaymentProcessing;
+
```
exit if Sphere is not turned on

bradymiller commented on sphere/process_revert_response.php:
```diff
@@ -0,0 +1,84 @@
+use OpenEMR\Common\Csrf\CsrfUtils;
+use OpenEMR\Core\Header;
+use OpenEMR\PaymentProcessing\PaymentProcessing;
+use OpenEMR\PaymentProcessing\Sphere\SphereRevert;
```
exit if Sphere is not turned on

bradymiller commented on sphere/token.php:
```diff
@@ -0,0 +1,52 @@
+use OpenEMR\Common\Acl\AclMain;
+use OpenEMR\Common\Csrf\CsrfUtils;
+use OpenEMR\PaymentProcessing\PaymentProcessing;
+use OpenEMR\PaymentProcessing\Sphere\SphereRevert;
```
exit if Sphere is not turned on

bradymiller commented:
This is now working and ready for review. I updated the initial PR message with details and screenshots.

@adunsulag , Converted to vanilla js and followed most other js recommendations. Pretty much logging everything (including unsuccessful transactions). Regarding the issue of bad actor modifying javascript, if a correctly authenticated user/patient decides to muck around with things by modifying javascript and directly making get/post requests then as with other places in OpenEMR things can be messed which would be logged (note that in the patient portal, changes require verification by a user in OpenEMR core), however, they will not be able to acquire sensitive info such as credit card numbers since this is done off of OpenEMR's server. Also the processing of callback requests from Sphere will require an open, active session and custom csrf token matching, so as above a correctly authenticated user/patient would need to decide to muck around with things that would be logged; as an aside the mechanism for the void/credit uses a more detailed sphere toolset where we get a token from sphere on backend, then direct to sphere with the token and get params (which need to match params sent to collect token), we then check hmac hash of the callback query from sphere (on the backend), and if that matches we then send a request to complete the transaction (on the backend).

bradymiller commented:
Working on wiki documentation here:
https://www.open-emr.org/wiki/index.php/Sphere_Payment_Processing_Module

Still working out details with Sphere on how accounts will be created etc. Then will get out a 6.0.0 patch (patch 4) with this feature :)

sjpadgett commented:
Nice work @bradymiller

sunsetsystems commented:
Thanks Brady, looks great!

stephenwaite commented:
![southpark_cheering](https://user-images.githubusercontent.com/5618372/148868376-44f67ede-d1a1-4260-bdfd-92bc765679ce.gif)
